doc/libfoa++.3
.TH "LIBFOA++" "3" "2009-03-28" "Nowise Systems" ""
\" This manual page is under GNU general public license (GPL).
\" Copyright (C) 2009-2019 Anders Lövgren
.SH NAME
C\+\+ library implementing FOA - Fast Object and Array encoding/decoding

.SH SYNOPSIS
\fB#include <libfoa\+\+.hpp>\fR

.SH DESCRIPTION
.PP
This C\+\+ library implements version 1.0 of the FOA specification. The spec 
can be found here: https://nowise.se/oss/libfoa/spec
.PP
FOA is designed to be used as an alternative to XML for web services, as an 
building block for object oriented network protocols or as an generic object 
serializing method (i.e. write objects in C\+\+ and read them back in Java).
.PP
The memory footprint of the library is small because it will never use more 
memory than required to hold the longest single line of data to decode.

.SH TYPES
This section briefly describes the FOA encoding. Each encoded line contains
either data (could be an array entry or an object member) or an special char. Each 
special char or data item must be followed by a newline (\\n) character. The 
special chars are 
.B '(' 
(begin object),
.B '[' 
(begin array),
.B ']' 
(end array) and
.B ')' 
(end object).
.PP
Each data, array or object (type) can also be named by 'name = type'. Arrays and 
objects can be unlimited nested in each other.
.PP
.B Syntax:
.PP
.br 
item := (name '=' data | data) + '\\n'
.br 
data := text | spec
.br 
name := text - '='
.br 
text := any char
.br 
spec := ([])
.br 
.PP
.B Example:
.PP
This is an example of how an array (named arr) containing two structures
(persons with named members) will be encoded:
.PP
arr = [
.br
(
.br
name = Albert Einstein
.br
born = March 14, 1879
.br
)
.br
(
.br
name = Isaac Newton
.br
born = December 25, 1642
.br
)
.br
]
.PP
.B Escape:
.PP
If escaping is turned on (the default), then special chars will be escaped when encoding
and unescaped when decoding (they are mapped to HTTP encoded equivalents).
This is all transparent to the library user. The escaping can be turned off
thru the \fBoption()\fR memeber function to improve performance.

.SH EXAMPLE
This simple example shows how to write an array of objects to std::cout. In
real world code, the output stream would be either an string stream (doing 
output buffering), an file stream or an tcp socket stream wrapper.
.RS
.nf

#include <iostream>
#include <libfoa++.hpp>

struct person
{
	const char *name;
	int age;
};

int main()
{
	person persons[] = {
		{ "adam", 35 },
		{ "bertil", 41 },
		{ "ceasar", 71 },
		{ 0, 0 }
	};
	
	foa::encoder enc(std::cout);
	
	enc.write(foa::start_array);
	for(person *pp = persons; pp->name; ++pp) {
		enc.write(foa::start_object);
		enc.write("name", pp->name);
		enc.write("age", pp->age);
		enc.write(foa::end_object);
	}
	enc.write(foa::end_array);
	
	return 0;
}
.RE
.fi

.SH NOTES
Full length examples of encoding/decoding and more documentation of the FOA
format is bundled with the source code.

.SH BUGS
Perhaps, send an bug report to andlov@nowise.se if you find one.

.SH AUTHOR
Anders Lövgren


NEWS
Version 1.0.2 (2019-01-29):

        * Migrated to GIT.
        
        * New author, bug report address and homepage URL.

	* Added shared representation of input buffer.
	
Version 1.0.1 (2009-03-29):

	* Support for building libfoa++ with Visual Studio 2003 added.

Version 1.0.0 (2009-03-28):

	* First release (Linux only).



src/libfoa++.cpp
// C++ library implementation of FOA - Fast Object and Array encoding.
// Copyright (C) 2009-2019 Anders Lövgren
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#ifdef HAVE_STRING_H
# include <string.h>
#elif HAVE_STRINGS_H
# include <strings.h>
#endif

#include "libfoa++.hpp"
#include "internal.hpp"

#ifndef HAVE_ISBLANK
int isblank(int c)
{
	return c == ' ' || c == '\t';
}
#endif

namespace foa {

	// 
	// Constructor for the shared parse buffer.
	// 
	parse_data::buff_rep::buff_rep() 
		: buffer(0), size(0), ppos(0), external(false), ref(0)
	{
	}
	
	// 
	// Only delete internal allocated buffer.
	// 
	parse_data::buff_rep::~buff_rep()
	{
		if(!external) {
			delete [] buffer;
		}
	}
	
	parse_data::scan_rep::scan_rep()
		: start(0), end(0), line(0)
	{
	}
	
	// 
	// Construct the parse data object.
	// 
	parse_data::parse_data() 
	{
		rep = new buff_rep;
	}

	// 
	// The copy contructor should share the parse buffer representation.
	// 
	parse_data::parse_data(const parse_data &data)
	{
		clone(data);
		rep->ref++;
	}
	
	// 
	// Delete the buffer only when buffer reference count drops below zero.
	// 
	parse_data::~parse_data()
	{
		if(--rep->ref < 0) {
			delete rep;
		}
	}

	// 
	// Increment buffer reference count on this object when assigned from
	// another object. We should *not* increment the refernce count here!
	// 
	void parse_data::operator=(const parse_data &data) 
	{
		clone(data);
	}
	
	// 
	// Reset the parse data. This method should be called prior to set
	// a new stream or buffer to decode.
	// 
	void parse_data::reset(bool external)
	{
		if(external != rep->external) {
			rep->buffer = 0;
		}
		rep->external = external;
		pos.line = pos.start = pos.end = rep->ppos = 0;
	}
	
	// 
	// Resize the buffer.
	// 
	void parse_data::resize(size_t size)
	{
		char *buffer = new char[size];
		if(rep->buffer) {
			memcpy(buffer, rep->buffer, rep->size);
			delete [] rep->buffer;
		}
		rep->size = size;
		rep->buffer = buffer;
	}
	
	// 
        // Move undecoded data to beginning of array. This method should be
	// called when no more entities where found in current buffer, but
	// before the buffer is filled again.
	// 
	void parse_data::move_data()
	{
		size_t length = rep->ppos - pos.end;
		memmove(rep->buffer + pos.end, rep->buffer, length);
		rep->ppos = length - 1;
		pos.start = pos.end = 0;
	}

	// 
	// Creates the object using default sizes for initial buffer 
	// size, reallocation chunk size and maximum buffer size.
	// 
	memory_strategy::memory_strategy()
	{
		init = INIT_SIZE;
		step = STEP_SIZE;
		max  = MAX_SIZE;
	}

	// 
	// Creates the object using the requsted sizes.
	// 
	memory_strategy::memory_strategy(size_t init, size_t step, size_t max)
	{
		this->init = init;
		this->step = step;
		this->max  = max;
	}
	
	// 
	// Creates the object using default sizes for initial buffer size 
	// and reallocation chunk size, but with no maximum buffer size.
	// 
	memory_strategy::memory_strategy(size_t init, size_t step)
	{
		this->init = init;
		this->step = step;
		this->max  = UNLIMITED;
	}
	
	// 
	// Replace all occurances of 'find' with 'replace' within the string 'str'.
	// 
	std::string escape::replace(std::string str, 
				    std::string::size_type pos,
				    const char *find, 
				    const char *replace)
	{
		size_t len = strlen(find);
		
		while((pos = str.find(find, pos)) != std::string::npos) {
			str = str.replace(pos, len, replace);
		}
		
		return str;
	}

	// 
	// Escape all special chars in str.
	// 
	std::string & escape::encode(std::string &str)
	{
		std::string::size_type pos = str.find_first_of("([])=");
		
		if(pos != std::string::npos) {
			str = replace(str, pos, "(", "%28");
			str = replace(str, pos, "[", "%5B");
			str = replace(str, pos, "]", "%5D");
			str = replace(str, pos, ")", "%29");
			str = replace(str, pos, "=", "%3D");
		}
		return str;
	}	

	// 
	// Unescape all special chars in str.
	// 
	std::string & escape::decode(std::string &str)
	{
		std::string::size_type pos = str.find('%');
		
		if(pos != std::string::npos) {
			str = replace(str, pos, "%28", "(");
			str = replace(str, pos, "%5B", "[");
			str = replace(str, pos, "%29", "]");
			str = replace(str, pos, "%5D", ")");
			str = replace(str, pos, "%3D", "=");
		}
		return str;
	}
}
